[AArch64][GCS][LLD] Introduce -zgcs-report-dynamic Command Line Option

Stylie777 opened the pull request:
When GCS was introduced to LLD, the gcs-report option allowed for a user to gain information relating to if their relocatable objects supported the feature. For an executable or shared-library to support GCS, all relocatable objects must declare that they support GCS.

The gcs-report checks were only done on relocatable object files, however for a program to enable GCS, the executable and all shared libraries that it loads must enable GCS. gcs-report-dynamic enables checks to be performed on all shared objects loaded by LLD, and in cases where GCS is not supported, a warning or error will be emitted.

It should be noted that only shared files directly passed to LLD are checked for GCS support. Files that are noted in the `DT_NEEDED` tags are assumed to have had their GCS support checked when they were created.

The behaviour of the -zgcs-dynamic-report option matches that of GNU ld. The behaviour is as follows unless the user explicitly sets the value:
* -zgcs-report=warning or -zgcs-report=error implies
-zgcs-report-dynamic=warning.

This approach avoids inheriting an error level if the user wishes to continue building a module without rebuilding all the shared libraries. The same approach was taken for the GNU ld linker, so behaviour is identical across the toolchains.

This implementation matches the error message and command line interface used within the GNU ld Linker. See here:
https://github.com/bminor/binutils-gdb/commit/724a8341f6491283cf90038260c83a454b7268ef

To support this option being introduced, two other changes are included as part of this PR. The first converts the -zgcs-report option to utilise an Enum, opposed to StringRef values. This enables easier tracking of the value the user defines when inheriting the value for the gas-report-dynamic option. The second is to parse the Dynamic Objects program headers to locate the GNU Attribute flag that shows GCS is supported. This is needed so, when using the gcs-report-dynamic option, LLD can correctly determine if a dynamic object supports GCS.

Stylie777 committed:
[AArch64][GCS][LLD] Introduce -z gcs-report-dynamic option

When GCS was introduced to LLD, the gcs-report option allowed
for a user to gain information relating to if their relocatable objects
supported the feature. For an executable or shared-library to support
GCS, all relocatable objects must declare that they support GCS.

The gcs-report checks were only done on relocatable object files,
however for a program to enable GCS, the executable and all
shared libraries that it loads must enable GCS.
gcs-report-dynamic enables checks to be performed on all shared
objects loaded by LLD, and in cases where GCS is not supported,
a warning or error will be emitted.

It should be noted that only shared files directly passed to LLD
are checked for GCS support. Files that are noted in the `DT_NEEDED`
tags are assumed to have had their GCS support checked when they
were created.

The behaviour of the -zgcs-dynamic-report option matches that of
GNU ld. The behaviour is as follows unless the user explicitly
sets the value:
* -zgcs-report=warning or -zgcs-report=error implies
-zgcs-report-dynamic=warning.

This approach avoids inheriting an error level if the
user wishes to continue building a module without rebuilding all
the shared libraries. The same approach was taken for the GNU ld
linker, so behaviour is identical across the toolchains.

This implementation matches the error message and command
line interface used within the GNU ld Linker. See here:
https://inbox.sourceware.org/binutils/20241206153746.3760179-2-matthieu.longo@arm.com/

To enable the checking of the GNU GCS Attributes when linking together
an application, LLD needs to be able to parse a Shared Files program
headers. This will enable a new option, allowing the user to output
diagnostic information relating to if there are shared files being
used that do not support the GCS extension.

To define if a shared file support GCS, the GNU GCS Attribute will
be stored within the Program Header, so this is parsed and, if
found, the `andFeatures` value updated to reflect the support level.

Stylie777 committed:
[AArch64][GCS] Convert -zgcs-report processing to use an Enum

To enable easier processing of inheritance for the -zgcs-report-dynamic
option, the -zgcs-report option has been converted to use an Enum value
to define its setting rather than a StringRef value.

This adds the Enum class for defining the three options, None, Warning
and Error, along with the appropirate methods for processing the users
input. To enable the error messages, a new lambda function has been
added. This processes the Enum value and passes the appropriate string
into the `report` lambda function for the error message. Error messages
have been updated where required.

smithp35 commented on lld/ELF/InputFiles.cpp:
```diff
@@ -1418,6 +1425,24 @@ std::vector<uint32_t> SharedFile::parseVerneed(const ELFFile<ELFT> &obj,
+  if (numElfPhdrs == 0)
+    return 0;
+  uint64_t attributes = 0;
+  for (unsigned i = 0; i < numElfPhdrs; i++)
```
I don't think that's how the PT_GNU_PROPERTY program header works.
The `p_flags` is generic ELF (see p_flags table in https://www.sco.com/developers/gabi/latest/ch5.pheader.html)

The p_offset and p_filesz give the location of the contents of the `.note.gnu.property section` which has the same structure as if it was in a relocatable object. You'll need to parse that contents of the file to look for GCS.

You should be able to reuse the code for parsing the notes section from a relocatable object.

If that code is too specific to an ELF section, then you could search the shared-libraries section header table. Strictly speaking this doesn't have to be there as it can be stripped, however LLD already seems to depend on it being present for other parts of parsing.

FWIW the Android Bionic Linker file https://android.googlesource.com/platform//bionic/+/16c23883bd707cb7b9827c44211a6cc5879e5032/linker/linker_note_gnu_property.cpp is what the Dynamic Linker does to process the PT_GNU_PROPERTY program header.

smithp35 commented on lld/test/ELF/aarch64-feature-gcs.s:
```diff
@@ -49,11 +49,37 @@
 # REPORT-WARN: warning: func2.o: -z gcs-report: file does not have GNU_PROPERTY_AARCH64_FEATURE_1_GCS property
 # REPORT-ERROR: error: func3.o: -z gcs-report: file does not have GNU_PROPERTY_AARCH64_FEATURE_1_GCS property
```
Have you got any tests where a shared library has the GNU property set? I can't immediately see one in the list below. That should pick up any problems in the parsing code.

smithp35 replied:
Yes, that is correct. You can tell by running `llvm-readelf --notes`
```
Displaying notes found in: .note.gnu.property
  Owner                Data size        Description
  GNU                  0x00000010       NT_GNU_PROPERTY_TYPE_0 (property note)
    Properties:    aarch64 feature: GCS
```

Stylie777 replied:
Infact, there is already a test to check this on line 18 and line 22
```
# RUN: llvm-readelf -n force-gcs.so | FileCheck --check-prefix GCS %s
...
# GCS: Properties:    aarch64 feature: GCS
```

MaskRay commented on lld/test/ELF/aarch64-feature-gcs.s:
```diff
@@ -49,11 +49,37 @@
 # INVALID: error: unknown -z gcs= value: nonsense
 
+## An invalid gcs option should give an error
+# RUN: not ld.lld func1-gcs.o func2-gcs.o func3-gcs.o -z gcs-report=nonsense 2>&1 | FileCheck --check-prefix=INVALID-GCS-REPORT %s
```
you can place `-z gcs-report=nonsense -z gcs-report-dynamic=nonsense` on the same RUN (to minimize lld invocations) and use just one .o file

MaskRay commented on lld/ELF/Driver.cpp:
```diff
@@ -569,6 +576,58 @@ static GcsPolicy getZGcs(Ctx &ctx, opt::InputArgList &args) {
+    // gcs-report, we want to inherit that value for gcs-report-dynamic. This is
+    // capped at a warning to ensure a users module can still build, while
+    // providing information relating to if a dynamic object supports GCS.
+    if (kv.first == "gcs-report" &&
```
Probably 

* Combine getZGcsReport and getZGcsReportDynamic
* Keep the last values of gcs-report and gcs-report-dynamic, respectively, in different variables.

Then one loop suffices.

Stylie777 reviewed (commented):
Thanks @MaskRay and @smithp35 for your reviews. I have responded and made improvements based on your comments.

Thanks!

MaskRay commented on lld/ELF/InputFiles.h:
```diff
@@ -224,8 +228,10 @@ class ELFFileBase : public InputFile {
   StringRef stringTable;
   const void *elfShdrs = nullptr;
   const void *elfSyms = nullptr;
+  const void *elfPhdrs = nullptr;
```
ELFFileBase is used by both .o and .so

move phdrs (image file only) to SharedFile and initialize them in SharedFile::parse

smithp35 commented on lld/ELF/InputFiles.cpp:
```diff
@@ -1418,6 +1423,67 @@ std::vector<uint32_t> SharedFile::parseVerneed(const ELFFile<ELFT> &obj,
+// information. This uses a similar process to the readGnuProperty
+// function, but designed specifically for SharedFiles.
+template <typename ELFT>
+void SharedFile::parseGnuAttributes(const uint8_t *base,
```
I think `parseGnuAndFeatures` would be a better name than GnuAttributes as this is parsing GNU Properties for AndFeatures.

smithp35 commented on lld/ELF/InputFiles.cpp:
```diff
@@ -1418,6 +1423,67 @@ std::vector<uint32_t> SharedFile::parseVerneed(const ELFFile<ELFT> &obj,
+    if (note.getType() != NT_GNU_PROPERTY_TYPE_0 || note.getName() != "GNU")
+      continue;
+
+    // Read a body of a NOTE record, which consists of type-length-value fields.
```
Would it be possible to extract this part so that it is common with readGnuProperty?

I think at this stage both are reading from an `ArrayRef<uint8_t>`

smithp35 commented on lld/test/ELF/aarch64-feature-gcs.s:
```diff
@@ -49,10 +49,28 @@
 
+## gcs-report-dynamic should report any dynamic objects that does not have the gcs property. This also ensures the inhertance from gcs-report is working correctly.
+
+# RUN: ld.lld func1-gcs.o func3-gcs.o no-gcs.so force-gcs.so -z gcs-report=warning -z gcs=always 2>&1 | FileCheck --check-prefix=REPORT-WARN-DYNAMIC %s
```
Can we add a test case with only inputs that have the marking `force-gcs.so` and `-zgcs-report=error` and `-zgcs-report-dynamic=error`? We expect no error.

MaskRay replied:
lld runs quickly, but 10+ invocations are too much for a relatively minor feature. Can the tests be simplified?

smithp35 reviewed (commented):
Thanks for the update. I didn't get a chance to look in detail today. Will take a look tomorrow.

smithp35 commented on lld/ELF/InputFiles.cpp:
```diff
@@ -918,6 +918,60 @@ void ObjFile<ELFT>::initializeSections(bool ignoreComdats,
+      // If the file being parsed is a SharedFile, we cannot pass in
+      // the data variable as there is no InputSection to collect the
+      // data from. As such, these are ignored. They are needed either
+      // when loading a shared library oject.
```
IIUC the data->data() is the full contents of the notes section. Whereas desc->data() is the contents of just one desc. Instead of skipping both the errors you could do something like
```
ArrayRef<uint8_t> contents = data ? *data : desc;
```
Then use contents.data() below.

MaskRay replied:
The comment is a bit too verbose regarding the lld code convention.

Actually, I am not sure shared objects should parse GNU_PROPERTY_AARCH64_FEATURE_PAUTH. It's orthogonal to this PR and this PR should not introduce the behavior change

smithp35 commented on lld/ELF/InputFiles.cpp:
```diff
@@ -1418,6 +1439,38 @@ std::vector<uint32_t> SharedFile::parseVerneed(const ELFFile<ELFT> &obj,
+void SharedFile::parseGnuAndFeatures(const uint8_t *base,
+                                     const typename ELFT::PhdrRange headers,
+                                     const typename ELFT::Shdr *sHeader) {
+  if (numElfPhdrs == 0 || sHeader == nullptr)
```
The program header and the section header are describing the same part of the ELF file.

For a shared object there is always a program header but there may not be a section header table (if the shared-library has been stripped).

It looks like the code only uses sHeader to get the section alignment. You can get this from the p_align field of the program header, then you don't need the sHeader at all.

smithp35 commented on lld/ELF/InputFiles.cpp:
```diff
@@ -1418,6 +1439,38 @@ std::vector<uint32_t> SharedFile::parseVerneed(const ELFFile<ELFT> &obj,
+      continue;
+    const typename ELFT::Note note(
+        *reinterpret_cast<const typename ELFT::Nhdr *>(base +
+                                                       headers[i].p_vaddr));
```
Can you check if p_vaddr is the right field? I think it should be p_offset? For most shared-libraries these will be the same for the PT_GNU_PROPERTY, but it is not guaranteed.

The p_vaddr is the virtual address, the p_offset is the file offset within the ELF file. The p_vaddr is meant to be used after the file has been memory mapped into virtual memory, when operating on the file p_offset should be used.

Most shared libraries start at address 0x0 but this is not a requirement so I'd expect using `p_vaddr` to break if a different base address had been used.

See https://www.sco.com/developers/gabi/latest/ch5.pheader.html

smithp35 reviewed (commented):
Thanks for the update. I think the use of p_vaddr might not always work, the other parts are suggestions for improvement.

Stylie777 commented on lld/ELF/InputFiles.cpp:
```diff
@@ -918,6 +918,60 @@ void ObjFile<ELFT>::initializeSections(bool ignoreComdats,
+                                 ArrayRef<uint8_t> &desc, ELFFileBase *f,
+                                 const uint8_t *base,
+                                 ArrayRef<uint8_t> *data = nullptr,
+                                 StringRef sectionName = ".note.gnu.property") {
```
I did originally have this, however the `err` lambda function within `readGnuProperty` takes the sectionName from the section variable. I wanted to preserve this functionality, with the default being `.note.gnu.property` which is used when parsing a SharedFile as that does not have the InputSection data available where the name comes from.

I can revert this back to inlining the string, but this is the reason for that variable being there.

smithp35 replied:
In this case I don't think the extra code to handle an ELF Note section of type `NT_GNU_PROPERTY_TYPE_0` but with a non standard name is worth it. In theory tools could handle it but in practice it looks like they don't. For example LLD only processes the section if it has that specific name https://github.com/llvm/llvm-project/blob/main/lld/ELF/InputFiles.cpp#L1060

For reference: the specification (https://github.com/hjl-tools/linux-abi/wiki/linux-abi-draft.pdf) defines the name to be `.note.gnu.property`, although the name doesn't have to be significant as the type field can be used.

smithp35 reviewed (approved):
LGTM on my side. Please leave some time for @MaskRay to comment.

For reference, the use of p_vaddr in Bionic's dynamic loader makes sense. At that point Bionic has already mmapped the ELF file into memory at p_vaddr (+ some displacement as shared libraries are position independent) so it is operating on the in-memory representation of the ELF file. LLD is reading directly from the file.

Stylie777 commented:
Thanks @smithp35. And that makes sense as to why we need to use `p_offset` rather than `p_vaddr` in LLD.

MaskRay commented on lld/ELF/Driver.cpp:
```diff
@@ -1553,6 +1587,9 @@ static void readConfigs(Ctx &ctx, opt::InputArgList &args) {
   ctx.arg.zForceBti = hasZOption(args, "force-bti");
   ctx.arg.zForceIbt = hasZOption(args, "force-ibt");
   ctx.arg.zGcs = getZGcs(ctx, args);
+  ctx.arg.zGcsReport = getZGcsReport(ctx, args, /* isReportDynamic */ false);
```
The canonical spelling is `/*xxx=*/false` without spaces

MaskRay commented on lld/docs/ReleaseNotes.rst:
```diff
@@ -25,6 +25,11 @@ Non-comprehensive list of changes in this release
 
 ELF Improvements
 ----------------
+* For AArch64, support for the `-zgcs-report-dynamic`` option has been added. This will provide users with
```
double backticks for .rst files

MaskRay commented on lld/ELF/Config.h:
```diff
@@ -136,6 +136,26 @@ enum LtoKind : uint8_t {UnifiedThin, UnifiedRegular, Default};
 enum class GcsPolicy { Implicit, Never, Always };
 
+// For -z gcs-report= and -zgcs-report-dynamic
+struct GcsReportPolicy {
```
We don't need a struct that wraps the enum. Just define an `enum` with a free function that converts the enum to "warning"/"error".

MaskRay commented on lld/ELF/Driver.cpp:
```diff
@@ -574,6 +576,38 @@ static GcsPolicy getZGcs(Ctx &ctx, opt::InputArgList &args) {
 }
 
+static GcsReportPolicy
+getZGcsReport(Ctx &ctx, opt::InputArgList &args, bool isReportDynamic,
```
I think the code will be cleaner if we compute the two values together in this function.

MaskRay commented on lld/ELF/InputFiles.h:
```diff
@@ -360,10 +360,22 @@ class SharedFile : public ELFFileBase {
+  }
+
+protected:
+  const void *elfPhdrs = nullptr;
```
getELFPhdrs is only needed once in `parse`. Do not cache them. `getELFPhdrs` can be deleted as well.

Stylie777 reviewed (commented):
Thanks @MaskRay, please see my changes, as you requested.

MaskRay committed:
Simplify code

MaskRay commented on lld/ELF/InputFiles.cpp:
```diff
@@ -1418,6 +1438,37 @@ std::vector<uint32_t> SharedFile::parseVerneed(const ELFFile<ELFT> &obj,
+    return;
+  uint32_t featureAndType = ctx.arg.emachine == EM_AARCH64
+                                ? GNU_PROPERTY_AARCH64_FEATURE_1_AND
+                                : GNU_PROPERTY_X86_FEATURE_1_AND;
```
This PR should not change the behavior of x86. DSOs are ignored for `GNU_PROPERTY_X86_FEATURE_1_AND` and I think we should continue ignoring it.

MaskRay commented on lld/docs/ReleaseNotes.rst:
```diff
@@ -25,6 +25,11 @@ Non-comprehensive list of changes in this release
 
 ELF Improvements
 ----------------
+* For AArch64, support for the ``-zgcs-report-dynamic`` option has been added. This will provide users with
```
This is too verbose. Consider

For AArch64, added support for -zgcs-report-dynamic, enabling checks for GNU GCS Attribute Flags in Dynamic Objects when GCS is enabled. Inherits value from -zgcs-report (capped at warning level) unless user-defined, ensuring compatibility with GNU ld linker.

MaskRay commented on lld/ELF/Driver.cpp:
```diff
@@ -574,6 +576,41 @@ static GcsPolicy getZGcs(Ctx &ctx, opt::InputArgList &args) {
   return ret;
 }
 
+static void getZGcsReport(Ctx &ctx, opt::InputArgList &args) {
```
Created  #130715 to introduce an `enum class` for none/warning/error.

Stylie777 replied:
Once this is merged, I will update this so we can utilise that `enum class`

MaskRay commented on lld/ELF/Driver.cpp:
```diff
@@ -1629,7 +1667,6 @@ static void readConfigs(Ctx &ctx, opt::InputArgList &args) {
       std::make_pair("cet-report", &ctx.arg.zCetReport),
       std::make_pair("execute-only-report", &ctx.arg.zExecuteOnlyReport),
-      std::make_pair("gcs-report", &ctx.arg.zGcsReport),
       std::make_pair("pauth-report", &ctx.arg.zPauthReport)};
```
I think it's better to handle gcs-report and gcs-report-dynamic here. We can add a special case to track whether gcs-report-dynamic has been specified

MaskRay replied:
I disagree. The is significant duplication and code bloat due to none/warning/error parsing.
You could add
```
if (option.first == "gcs-report-dynamic")
  reportDynamicSpecified = true
```

then after the for loop,
```
  if (!reportDynamicDefined && ctx.arg.zGcsReport != ReportPolicy::None &&
      ctx.arg.zGcsReportDynamic == ReportPolicy::None)
    ctx.arg.zGcsReportDynamic = ReportPolicy::Warning;
```

MaskRay commented on lld/ELF/Driver.cpp:
```diff
@@ -2788,6 +2825,17 @@ static void redirectSymbols(Ctx &ctx, ArrayRef<WrappedSymbol> wrapped) {
     ctx.symtab->wrap(w.sym, w.real, w.wrap);
 }
 
+static StringRef gcsReportPolicytoString(GcsReportPolicy value) {
```
could be removed with #130715

Stylie777 committed:
Respond to Review Comments

Stylie777 committed:
Remove `gcsReportPolicytoString` function

This is no longer needed now this PR is using the new
`ReportPolicy` enum.

Stylie777 committed:
Fix release notes issue

Stylie777 committed:
Fix formatting in InputFiles.cpp

MaskRay commented on lld/ELF/InputFiles.cpp:
```diff
@@ -1418,6 +1434,36 @@ std::vector<uint32_t> SharedFile::parseVerneed(const ELFFile<ELFT> &obj,
+    // Read a body of a NOTE record, which consists of type-length-value fields.
+    ArrayRef<uint8_t> desc = note.getDesc(headers[i].p_align);
+    parseGnuPropertyNote<ELFT>(
+        ctx, *this, /*featureAndType*/ GNU_PROPERTY_AARCH64_FEATURE_1_AND, desc,
```
remove `/*featureAndType*/`. it's clear from the argument value (GNU_PROPERTY...)

If we need to name the argument, the canonical form is `/*xxx=*/value` (no space)

Stylie777 committed:
Update parsing of gcs-report and gcs-report-dynamic options

This removes the new function that was originally added, and
reinstates the parsing of the gcs-report option in its original
location. This also enables the use of gcs-report-dynamic in
the reports, with the GNU inheritance implemented.

MaskRay committed:
make code more conventional; simplify code; use just diag prefix in CHECK-NOT.
